src/interpolation_utils.jl

```diff
@@ -57,3 +57,68 @@ function munge_data(U::StridedMatrix, t::AbstractVector)
 
   return hcat(newUs...), newt
 end
+
+
+"""
+    bracketstrictlymontonic(v, x, guess; lt=<comparison>, by=<transform>, rev=false)
+
+Starting from an initial `guess` index, find indices `(lo, hi)` such that `v[lo] ≤ x ≤
+v[hi]` according to the specified order, assuming that `x` is actually within the range of
+values found in `v`.  If `x` is outside that range, either `lo` will be `firstindex(v)` or
+`hi` will be `lastindex(v)`.
+
+Note that the results will not typically satify `lo ≤ guess ≤ hi`.  If `x` is precisely
+equal to a value that is not unique in the input `v`, there is no guarantee that `(lo, hi)`
+will encompass *all* indices corresponding to that value.
+
+This algorithm is essentially an expanding binary search, which can be used as a precursor
+to `searchsorted` and related functions, which can take `lo` and `hi` as arguments.  The
+purpose of using this function first would be to accelerate convergence in those functions
+by using correlated `guess`es for repeated calls.  The best `guess` for the next call of
+this function would be the index returned by the previous call to `searchsorted`.
+
+See [`sort!`](@ref) for an explanation of the keyword arguments `by`, `lt` and `rev`.
+"""
+function bracketstrictlymontonic(
+  v::AbstractVector,
+  x,
+  guess::T,
+  o::Base.Order.Ordering
+)::NTuple{2,keytype(v)} where {T<:Integer}
+  bottom = firstindex(v)
+  top = lastindex(v)
+  if guess < bottom || guess > top
+    return bottom, top
+    # # NOTE: for cache efficiency in repeated calls, we avoid accessing the first and last elements of `v`
+    # # on each call to this function.  This should only result in significant slow downs for calls with
+    # # out-of-bounds values of `x` *and* bad `guess`es.
+    # elseif lt(o, x, v[bottom])
+    #     return bottom, bottom
+    # elseif lt(o, v[top], x)
+    #     return top, top
+  else
+    u = T(1)
+    lo, hi = guess, min(guess + u, top)
+    @inbounds if Base.Order.lt(o, x, v[lo])
+      while lo > bottom && Base.Order.lt(o, x, v[lo])
+        lo, hi = max(bottom, lo - u), lo
+        u += u
+      end
+    else
+      while hi < top && !Base.Order.lt(o, x, v[hi])
+        lo, hi = hi, min(top, hi + u)
+        u += u
+      end
+    end
+  end
+  return lo, hi
+end
+
+for s in [:searchsortedfirst, :searchsortedlast]
+  @eval begin
+    function Base.$s(v::AbstractVector, x, guess)

```

Comment:
Extending `searchsortedfirst` is type piracy. We should just make a new function.

Reply:
Renamed to `searchsortedfirstcorrelated` (and `...last...`)